File: cyclic_rotation_2/opussf/cr2.py
#!/usr/bin/env python

def solution( listIn, offset, swapLength ):
	""" reverse only a portion of listIn, starting at offset, and reversing swapLength items
	"""
	# find the slice to reverse
	outList = listIn[offset:offset+swapLength]
	# reverse it
	outList.reverse()
	# reconstruct the list
	outList = listIn[:offset]+outList+listIn[offset+swapLength:]
	# return the list
	return outList

if __name__=="__main__":
	import unittest

	class testCyclicRotation2( unittest.TestCase ):
		def setUp( self ):
			pass
		def test01( self ):
			self.assertEquals( solution( [ 3, 8, 9, 7, 6 ], 1, 3 ), [ 3, 7, 9, 8, 6 ] )
		def test02( self ):
			self.assertEquals( solution( [ 3, 8, 9, 7, 6 ], 0, 3 ), [ 9, 8, 3, 7, 6 ] )
		def test03( self ):
			self.assertEquals( solution( [ 3, 8, 9, 7, 6 ], 0, 5 ), [ 6, 7, 9, 8, 3 ] )
		def test04( self ):  # Bonus!  it goes 'backwards' too
			self.assertEquals( solution( [ 3, 8, 9, 7, 6 ], 1, 4 ), [ 3, 6, 7, 9, 8 ] )
		def test05( self ):
			self.assertEquals( solution( solution( [ 3, 8, 9, 7, 6 ], 1, 2 ), 1, 4 ), [ 3, 6, 7, 8, 9 ] )

	unittest.main()

File: cyclic_rotation_2/opussf/cr2.lua
#!/usr/bin/env lua

function solution( tableIn, offset, swapLength )
	newTable = {}

	-- copy non-reversed
	for lcv = 1, offset do  -- offset = 0 == no initial data
		table.insert( newTable, tableIn[lcv] )
	end
	-- copy reversed section
	for lcv = offset+swapLength, offset+1, -1 do -- offset+1 to go to 1, not 0.
		table.insert( newTable, tableIn[lcv] )
	end
	-- copy non-reversed
	for lcv = offset+swapLength+1, #tableIn do
		table.insert( newTable, tableIn[lcv] )
	end
--[[
	for lcv=1, #newTable do
		print( string.format("%02i: %i", lcv, newTable[lcv] ) )
	end
]]
	return newTable
end
